src/Common/src/TypeSystem/Interop/IL/Marshaller.cs

```diff
@@ -1681,9 +1747,17 @@ protected override void TransformManagedToNative(ILCodeStream codeStream)
         }
         protected override void AllocNativeToManaged(ILCodeStream codeStream)
         {
-            codeStream.Emit(ILOpcode.call, _ilCodeStreams.Emitter.NewToken(
+            var emitter = _ilCodeStreams.Emitter;
+            var lNull = emitter.NewCodeLabel();
+
+            // Check for null array
+            LoadNativeValue(codeStream);
+            codeStream.Emit(ILOpcode.brfalse, lNull);
+
+            codeStream.Emit(ILOpcode.newobj, emitter.NewToken(
```

Comment:
You need to emit the right capacity for byref case. See https://github.com/dotnet/coreclr/blob/master/src/vm/ilmarshalers.cpp#L828-L842